### PR TITLE
Add code coverage reports via Coveralls

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

This PR:
- Updates the existing but unused `m4/pdns_enable_coverage.m4` to use modern compiler options, and hook it up to our `configure.ac` scripts
- Makes sure that coverage data is properly recorded when our products exit, even if they call `_exit()` for example
- Generates coverage data in our CI as part of the `build-and-test-all` workflow, so that it is executed for every commit and pull request
- Documents the process in `CODE_COVERAGE.md`

We still need to grant access to our repository to Coveralls for this to work, but I wanted to get feedback first.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
It looks like we already got a report, even though it's not yet officially enabled/allowed on GH's side: https://coveralls.io/github/PowerDNS/pdns

File: CODE_COVERAGE.md
Comment (fredmorcos) [inline CODE_COVERAGE.md:6]:
```suggestion
Code coverage is generated during our Continuous Integration tests, for every pull request. In addition to the dashboard on Coveralls' website, a summary is posted on pull requests.
```

File: CODE_COVERAGE.md
Comment (fredmorcos) [inline CODE_COVERAGE.md:4]:
```suggestion
PowerDNS uses [coveralls](https://coveralls.io/) to generate code coverage reports from our Continuous Integration tests. The resulting analysis can then be consulted [online](https://coveralls.io/github/PowerDNS/pdns), and gives insight into which parts of the code are automatically tested.
```

File: CODE_COVERAGE.md
Comment (fredmorcos) [inline CODE_COVERAGE.md:76]:
```suggestion
The way our code coverage report is generated does not currently handle the different authoritative server tools (that end up in the `pdns-tools` package) very well. Consequently the coverage report for these tools, and the related code parts, is not accurate.
```

**Review by fredmorcos — APPROVED** [review]
Awesome stuff. I can't reliably review anything under `.github/` though.

File: pdns/coverage.cc
Comment (fredmorcos) [reply pdns/coverage.cc:28]:
```suggestion
  // NOLINTNEXTLINE(bugprone-reserved-identifier): not ours
```

File: pdns/coverage.cc
Comment (fredmorcos) [reply pdns/coverage.cc:31]:
```suggestion
  // NOLINTNEXTLINE(bugprone-reserved-identifier): not ours
```

File: CODE_COVERAGE.md
Comment (rgacogne) [inline CODE_COVERAGE.md:41]: The link is wrong here

**Comment by omoerbeek** [comment]
From te report:
```
 30435 of 84400 branches covered (0.0%)

Branch coverage included in aggregate %.
```
How should I interpret these numbers?

Drilling down to the source files does not show me relevant info, I see a lot of "source not availabe" ot "The commit SHA "ffd07a92b8a9a9e262ab6022c2fd88e5b26c6d95" was not found in your repository, so the file cannot be loaded. This may be because you posted from a local development environment, or your CI created an ephemeral commit."

**Comment by rgacogne** [comment]
I can see the error as well, but it wasn't there ten days ago when I submitted the PR, so my guess is something expired. We are in a weird state with regard to Coveralls right now because I did not want to grant any access to them until after getting approval from the team, so this might explain why we are seeing these issues. I can try force-pushing if you want to look at a (hopefully) correct report?

**Comment by omoerbeek** [comment]
yes please

**Comment by rgacogne** [comment]
So I did force-push, but it turns out that I can see the content of the files only if I'm logged into Coveralls. Given that this PR comes from a repository on my branch, and that I have granted access to my repository to Coveralls, this might be related. I'm not sure how to fix this without granting access to Coveralls to the PowerDNS/pdns repository, shall I do that?

**Comment by rgacogne** [comment]
I have granted Coveralls access to the PowerDNS/pdns repository, and I have claimed ownership of the report on the Coveralls website. It looks better for me now, for example https://coveralls.io/builds/62875508/source?filename=pdns%2Fauth-querycache.cc works properly even if I'm not logged into the Coveralls website.

**Comment by omoerbeek** [comment]
Yes, this works for me too. I think this should be merged. We can always improve things later if needed.

**Review by omoerbeek — APPROVED** [review]

**Comment by rgacogne** [comment]
@Habbie do you want to have a look before I merge this PR?

File: CODE_COVERAGE.md
Comment (Habbie) [inline CODE_COVERAGE.md:30]:
```suggestion
### Source Based
```

File: pdns/coverage.cc
Comment (Habbie) [inline pdns/coverage.cc:29]: `#ifdef COVERAGE` ?

File: pdns/coverage.cc
Comment (Habbie) [inline pdns/coverage.cc:44]: should this be inside -both- ifdefs?

**Review by Habbie — COMMENTED** [review]
Seems fine! One typo, couple of questions

File: pdns/coverage.cc
Comment (Habbie) [reply pdns/coverage.cc:29]: it is!

File: pdns/coverage.cc
Comment (rgacogne) [reply pdns/coverage.cc:44]: It's a good point, we might not want to call `gcov_dump()` when source-based coverage is enabled because it's a useless. On the other hand, it's a no-op. I'll remove it.

**Comment by rgacogne** [comment]
I pushed two new commits, fixing the typo (plus another one!) and no longer calling `__gcov_dump` when source-based coverage is used. I'll merge this PR once the tests are green.

**Review by Habbie — APPROVED** [review]

**Comment by rgacogne** [comment]
Rebased on master.

**Pull request merged**
